### PR TITLE
ARROW-14941 [R] Implement Duration R6 class and bindings for lubridate::duration()

### PR DESCRIPTION
**Pull request by paleolimbot**
This PR implements the Duration type in the R bindings, mapped to "difftime", which is the equivalent base R type. Currently a draft PR because I'm unsure how to implement the `Extend()` method of the converter:

https://github.com/apache/arrow/blob/7240420a38d4b12317d54fc5195ddcaa5f35cf61/r/src/r_to_arrow.cpp#L824-L826

I've also copy-and-pasted the converter for the TIME32/hms object conversion because I was unsure how to subclass it to avoid the repetition in the moment.

The gist of the PR is to allow this (which currently fails):

``` r
library(arrow, warn.conflicts = FALSE)
Array$create(as.difftime(123, units = "secs"))
#> Error: NotImplemented: Extend
```

<sup>Created on 2021-12-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

File: r/src/r_to_arrow.cpp [hunk 1]
```diff
@@ -580,33 +583,37 @@ int64_t get_TimeUnit_multiplier(TimeUnit::type unit) {
...
 template <typename T>
 class RPrimitiveConverter<T, enable_if_t<is_time_type<T>::value>>
     : public PrimitiveConverter<T, RConverter> {
  public:
   Status Extend(SEXP x, int64_t size, int64_t offset = 0) override {
     RETURN_NOT_OK(this->Reserve(size - offset));
     auto rtype = GetVectorType(x);
-    if (rtype != TIME) {
+    // This probably also not accept duration
+    if (rtype != TIME && rtype != DURATION) {
```
Comment (romainfrancois) [inline r/src/r_to_arrow.cpp:611]:
```suggestion
    if (rtype != TIME) {
```

I believe `DURATION` (i.e. <difftime>) should not be dealt with here.

File: r/src/r_to_arrow.cpp [hunk 2]
```diff
@@ -580,33 +583,37 @@ int64_t get_TimeUnit_multiplier(TimeUnit::type unit) {
...
+}
+
 template <typename T>
 class RPrimitiveConverter<T, enable_if_t<is_time_type<T>::value>>
     : public PrimitiveConverter<T, RConverter> {
  public:
   Status Extend(SEXP x, int64_t size, int64_t offset = 0) override {
     RETURN_NOT_OK(this->Reserve(size - offset));
     auto rtype = GetVectorType(x);
+    // This probably also not accept duration
```
Comment (romainfrancois) [inline r/src/r_to_arrow.cpp:610]: the comment is probably obsolete then

**Review by romainfrancois — APPROVED** [review]

**Comment by nealrichardson** [comment]
Thanks for this! Looks like there's one more entry in arrow.Rmd, showing the mapping of Arrow Duration type to R, the table below the one that was edited.

**Comment by jonkeane** [comment]
Thanks for catching that. I've made https://issues.apache.org/jira/browse/ARROW-15082

**Comment by thisisnic** [comment]
> I haven't implemented the `lubridate::duration()` translation because it seems out of scope for this set of changes but could also include it with this PR.

All good, though would you mind opening a new ticket for that?

**Pull request closed without merging**